ride/swap.ride

```diff
@@ -0,0 +1,196 @@
+
+let poolDisabled = "Pool disabled".throw()
+let poolNotExist = "Pool is not exist".throw()
+let slippageTooBig = "Slippage is too big".throw()
+
+func keyManagerPublicKey() = "%s__managerPublicKey"
+func keyPendingManagerPublicKey() = "%s__pendingManagerPublicKey"
+
+func fc() = {"%s__factoryContract"}
+let factoryContract = addressFromStringValue(getStringOrFail(fc()))
+
+func feeLiquidityContractAddress() = ["%s", "feeLiquidityContractAddress"].makeString(SEP)
+func feeLiquidity() = ["%d", "feeLiquidity"].makeString(SEP)
+func protocolFee() = ["%d", "protocolFee"].makeString(SEP)
+
+
+func asInt(val: Any) = {
+  match val {
+    case valInt: Int => valInt
+    case _ => throw("fail to cast into Int")
+  }
+}
+
+func dataMappingPoolAssets(internalAmountAssetStr: Int, internalPriceAssetStr: Int) = {
+  makeString(
+    ["%d%d",
+    internalAmountAssetStr.toString(),
+    internalPriceAssetStr.toString()],
+    SEP)
+}
+
+func managerPublicKeyOrUnit() = match keyManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func pendingManagerPublicKeyOrUnit() = match keyPendingManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func isManager(i: Invocation) = {
+  match managerPublicKeyOrUnit() {
+    case pk: ByteVector => i.callerPublicKey == pk
+    case _: Unit => i.caller == this
+  }
+}
+
+func mustManager(i: Invocation) = {
+  i.isManager() || "permission denied".throw()
+}
+
+@Callable(i)
+func constructor(
+  protocolFeeAmount: Int,
+  feeLiquidityAmount: Int,
+  feeLiquidityAddress: String,
+  fc: String
+  ) = {
+  strict check = i.mustManager()
+  [
+    StringEntry(feeLiquidityContractAddress(), feeLiquidityAddress),
+    IntegerEntry(feeLiquidity(), feeLiquidityAmount),
+    IntegerEntry(protocolFee(), protocolFeeAmount),
+    StringEntry(fc(), fc)
+  ]
+}
+
+
+
+@Callable(i)
+func swap(
+  amountOutMin: Int,
+  assetOut: String,
+  addressTo: String # add if there is not added
+) = {
+  let pmt = i.payments[0].value()
+  let assetIn = pmt.assetId.value()
+  let amountIn = pmt.amount.value()
+  let prFee = getInteger(protocolFee()).value()
+  let feeProtocolAmount = fraction(amountIn.toBigInt(), prFee.toBigInt(), feeScale)
+  let feeLiquidityAmount = fraction(amountIn.toBigInt(), prFee.toBigInt(), feeScale)
+
+  let amountInToPool = amountIn.toBigInt() - feeProtocolAmount
+  let cleanAmountIn = amountInToPool -  feeLiquidityAmount
+
+  strict checks = [
+    i.payments.size() == 1 || "exactly 1 payment are expected".throw()
+  ]
+
+  # getLpAssetFromPoolAssetsREADONLY
+  let lpNonReverese = invoke(
+    factoryContract,
+    "getLpAssetFromPoolAssetsREADONLY",
+    [assetIn, assetOut],
+    []
+  )
+
+  let lpReverse = invoke(
+    factoryContract,
+    "getLpAssetFromPoolAssetsREADONLY",
+    [assetOut, assetIn],
+    []
+  )
+  let (isReverse, pool) = if (lpNonReverese != unit) then {
```

Comment:
The `pool` variable have `Any` type
`match` or `exactAs` have to be used

Reply:
agree. corrected. now using exact